### ISSUE TITLE
Web playback of Full Tracks

### ISSUE THREAD

**Issue by func0der** [spotify/web-api#57]
Hey there,

I had a quick look at your Web API. It seems that there is the possiblity to get a 30 seconds preview of a song via a streaming url your a providing in the results.
But I could not find the 'whole song' streaming url in there. 

Did I miss something or is streaming your songs really not possible via this or any other API?

Thank you

**Comment by thelinmichael** [comment]
You didn't miss anything - retrieving links to full mp3s is not possible using the Web API. Playback of full tracks is possible using the Android and iOS SDKs though.

**Comment by func0der** [comment]
Why is this not possible with the Web API?

**Comment by thelinmichael** [comment]
It's not possible yet, we're looking into providing a way for developers to do playback of full-length tracks over the web. Exactly how hasn't been decided.

**Comment by func0der** [comment]
That is actually great news. Thank you for the insight :)

I hope this will be available soon.

**Comment by jonknyc** [comment]
Thanks for pointing me to this issue - it seems like you should be able to just wrap whatever powers play.spotify.com, including the requirement that the user be logged into Spotify, and that can be a good v1.

Can't say enough how bad I'd love to use this API!

**Comment by tom-parker** [comment]
Any ideas on when this might surface? Are you talking weeks or months when you mention short term?

**Comment by filfat** [comment]
+1

**Comment by GabeJacobs** [comment]
@thelinmichael Hey! I was wondering if there was any news on this. We would absolutely love to beta test any kind of full playback web functionality. Could be helpful on both ends so that we can report bugs, help solve issues.

**Comment by karlmonson** [comment]
+1, would love this functionality.

**Comment by thelinmichael** [comment]
No news unfortunately @GabeJacobs. Definitely see the benefits of having third-party developers doing beta testing.

**Comment by drsounds** [comment]
This a reason why Spotify Apps shouldnt havent closed until full web
streaming or play button js sdk are available. Now mobile developers gets
better support than desktop
Den 29 okt 2015 09:38 skrev "Michael Thelin" <notifications@github.com>:

**Comment by Kingdorian** [comment]
Is there any news if (and when) this is going to happen?

**Comment by philonor** [comment]
+1

**Comment by gnumdk** [comment]
Really strange, that hard to add authentification on an http server?

So, if I launch wireshark, I can get track urls and play them without any authentification?

**Comment by drsounds** [comment]
We havent seen any developer news since june 2015. Thats strange for a such
top notch company. Are you retiring Spotify desktop entirely? The video
feature announced will only be available for mobile?

onsdag 3 februari 2016 skrev Cédric Bellegarde <notifications@github.com>:

> Really strange, that hard to add authentification on an http server?
>
> So, if I launch wireshark, I can get track urls and play them without any
> authentification?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/web-api/issues/57#issuecomment-179191743>.
>


-- 
Best regards,
Alexander Forselius (Dr. Sounds)
Musician, and programmer.
http://dr-sounds.com

Github <http://www.github.com/krikelin>
Twitter <http://www.twitter.com/drsounds>
Spotify <http://open.spotify.com/artist/2FOROU2Fdxew72QmueWSUy>
<http://www.youtube.com/watch?v=pjdujamrcds>
LinkedIn <http://se.linkedin.com/pub/alexander-forselius/13/b11/500/>

**Comment by vincelwt** [comment]
+1, an ETA from Spotify would be great

**Comment by benpye** [comment]
With libspotify being discontinued, it would be great to get this capability in the web api, platforms exist outside of Android and iOS. It would also be nice to be able to do 320kbps streaming...

**Comment by rvoortman** [comment]
I feel like spotify is taking a crap on all developers, but also on users... Android and iOS are pretty well supported, but all other platforms are just barely, if at all, supported. Spotify, please provide information so developers and users know what to expect.

**Comment by RareNCool** [comment]
So is there any workaround to play full tracks on, say, a Windows 10 app? I tried repackaging the Android .jar files as .dll's, but there are some issues. Spotify really needs to get on this.

**Comment by vincelwt** [comment]
A workaround I found (not sure if it violate Spotify's conditions) is to
get tracks metadatas from the Spotify API,  and then get the audio from
YouTube. Doesn't works on 100% of the tracks tough.

Le mar. 15 mars 2016 17:18, Shawn Beachy <notifications@github.com> a
écrit :

> So is there any workaround to play full tracks on, say, a Windows 10 app?
> I tried repackaging the Android .jar files as .dll's, but there are some
> issues. Spotify really needs to get on this.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/spotify/web-api/issues/57#issuecomment-196904135
>
-- 

*--*

*Vincent L.*

**Comment by drsounds** [comment]
That workaround defintely violate the TOS but the previous one seems to be
interesting the repackaging of the java files into dll but I am sure it is
a violation too

tisdag 15 mars 2016 skrev Vincent Lwt <notifications@github.com>:

> A workaround I found (not sure if it violate Spotify's conditions) is to
> get tracks metadatas from the Spotify API, and then get the audio from
> YouTube. Doesn't works on 100% of the tracks tough.
>
> Le mar. 15 mars 2016 17:18, Shawn Beachy <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> a
> écrit :
>
> > So is there any workaround to play full tracks on, say, a Windows 10 app?
> > I tried repackaging the Android .jar files as .dll's, but there are some
> > issues. Spotify really needs to get on this.
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly or view it on GitHub:
> > https://github.com/spotify/web-api/issues/57#issuecomment-196904135
> >
> --
>
> *--*
>
> *Vincent L.*
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/spotify/web-api/issues/57#issuecomment-196920761
>


-- 
Best regards,
Alexander Forselius (Dr. Sounds)
Musician, and programmer.
http://dr-sounds.com

Github <http://www.github.com/krikelin>
Twitter <http://www.twitter.com/drsounds>
Spotify <http://open.spotify.com/artist/2FOROU2Fdxew72QmueWSUy>
<http://www.youtube.com/watch?v=pjdujamrcds>
LinkedIn <http://se.linkedin.com/pub/alexander-forselius/13/b11/500/>

**Comment by RareNCool** [comment]
@drsounds  It very well may be. However, not to casually pass off violation of TOS, but stuff like that *is* gonna end up happening if Spotify doesn't give us an official way.

**Comment by drsounds** [comment]
Yes and that might cause the entire api being shut down I am afraid

tisdag 15 mars 2016 skrev Shawn Beachy <notifications@github.com>:

> @drsounds <https://github.com/drsounds> It very well may be. However, not
> to casually pass off violation of TOS, but stuff like that *is* gonna end
> up happening if Spotify doesn't give us an official way.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/spotify/web-api/issues/57#issuecomment-196922519
>


-- 
Best regards,
Alexander Forselius (Dr. Sounds)
Musician, and programmer.
http://dr-sounds.com

Github <http://www.github.com/krikelin>
Twitter <http://www.twitter.com/drsounds>
Spotify <http://open.spotify.com/artist/2FOROU2Fdxew72QmueWSUy>
<http://www.youtube.com/watch?v=pjdujamrcds>
LinkedIn <http://se.linkedin.com/pub/alexander-forselius/13/b11/500/>

**Comment by jakemason** [comment]
Very disappointed that 30 second previews are all we can work with. Really hope this feature gets some love from the team soon!

**Comment by rtrompier** [comment]
+1 ...

**Comment by indiecore-fm** [comment]
This is a sad state of affairs here. Last.FM api is also going to the ********. We have been waiting on this feature for a long time. Can we can get any information? This is apparently something all the developers are wanting, so drop the play button and let us build clients!

**Comment by tansut** [comment]
It seems we can get all information for a track but cannot play it!

**Comment by jonknyc** [comment]
Here are some ideas for cool features that I'd love to build, if Spotify
had API playback support:

- playlists with jump-in / jump-out points (so tracks can start / end at a
certain point; super important for creating dance playlists)

- virtual jukebox, where guests at your party crowdsource the playlist

- ability to beat match / normalize volume between tracks

- "midnight song": this is for midnight on New Year's Eve.  You put Times
Square NYC on TV (muted), and want to get a song to hit the chorus just as
the ball drops at the stroke of midnight. (a great example is Rihanna "We
Found Love" - the build-up to 1:07 is epic; so gotta start the song at
11:58:53pm). (I've done this a few times with custom software - it's the
coolest thing ever) Anyway, it would be great to use the Spotify API to get
your midnight song to start at just the right moment so the beat drops at
the stroke of midnight.

**Comment by onreachable** [comment]
This feature would allow an infinite number of new concepts for music apps, I believe that Spotify devs would welcome that. However "the web" can't be considered a controlled environment just as mobile platforms are. For example, the current web player has dozens off apps that records the songs and save them as mp3 files, so it wouldn't make much sense to publish an API that makes this even easier.

We would do better by giving ideas of how to protect Spotify's content from illegitimate use instead of complaining why this feature isn't available yet.

AFAIK, Netflix is successfully dealing with this by working within W3C in some drafts that will help browsers to protect media playback in general... are you guys considering something similar?!

@thelinmichael if an ETA can't be provided, we would like to know at least if currently this a low priority feature or not.

Thanks!

**Comment by jordanranson** [comment]
+1 would like to see this feature ASAP.

**Comment by crwgregory** [comment]
I think what the problem for Spotify with this sort of feature is that they are afraid of violating someones personal privacy. Case in point, the web play button has no way to automatically play, the user has to click the play button. No one wants a website to play anything when it loads. If that is the concern, a possible solution might be:

When the user grants access to their profile to a web application, there could be a scope of: 'Allow this application to control Spotify Desktop app'. Getting consent from the user.

Using another endpoint 'clientPlay' (or something) you would pass the Oauth token and Spotify would determine authenticity, then it would just funnel the request into what open.spotify.com already does.

I doubt I'm the first one to mention this, but thought I would put it out there.

**Comment by levi-cole** [comment]
Just a thought... How do Sonos get around this? Do they have permissions that the rest of us don't? 🤔

**Comment by everdev** [comment]
@jonknyc I came along this thread researching the API to make a party jukebox app like you mentioned. If you're interested in collaborating, feel free to send me an Email (it's in my profile).

**Comment by brijeshrakholia** [comment]
Is this feature out yet?

**Comment by drsounds** [comment]
My educated guess

Spotify has got in trouble with some lawsuits and also needs to be
substainable, the pressure to reduce the loss and make money has increased
so I think they are careful to introduce this unless times becomes better.
web streaming could need negotiations with rightholders and I think they
are not willing to do this right now. There is a difference between
allowing web and mobile streaming and it is that a web streaming service
would increase the risk that more people rips music from spotify. So I
propose a JS api to the play button instead which can control spotify from
the browser although the desktop client will do the playback. But this can
be tricky too for business reason since they dont want people to create
services like eternify was if programmatic access to playback would be
possible.

A spotify connect api or a js api for play button would be very handy. But
before this can be done I think some investigation must be done. And
Spotify also become more substainable.

It seems like the developer platform has slowed down in its development and
they focus more on increasing revenue streams with more branding options
such branded playlists.

fredag 10 juni 2016 skrev Brijesh Rakholia <notifications@github.com>:

> Is this feature out yet?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spotify/web-api/issues/57#issuecomment-225242204>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AE3z1yfHiQrjA9JHitVSHdRMbZnYBGpQks5qKZw-gaJpZM4EqAUH>
> .
>


-- 
Best regards,
Alexander Forselius (Dr. Sounds)
Musician, and programmer.
http://dr-sounds.com

Github <http://www.github.com/krikelin>
Twitter <http://www.twitter.com/drsounds>
Spotify <http://open.spotify.com/artist/2FOROU2Fdxew72QmueWSUy>
<http://www.youtube.com/watch?v=pjdujamrcds>
LinkedIn <http://se.linkedin.com/pub/alexander-forselius/13/b11/500/>

**Comment by Ofekmeister** [comment]
Any updates?

**Comment by arn95** [comment]
People are saying the Android and iOS SDKs are better... Are you guys serious? Have you even had a look at their limited exposure on the player functions and the foreground service the player runs on at least on Android? I have been working on my project for a month until I realized that Spotify, a music company, has no way to let you play and control music properly. It's a shame. Why make any of the APIs available when all you can do is log in and browse? ...

**Comment by gdaunton** [comment]
+1

**Comment by crwgregory** [comment]
From a business stand point, if you had many third party apps playing your music without regulation, that would lose the company money without the ability to manage advertisements or accounts. With one payed account hooked up to a Spotify API, the third party app could control the player at will, potentially bringing music to hundreds or thousands with one payed account.

So how do you regulate the use of a feature like this?

You create API credits, and playing tracks cost credits. The more popular the app, the more credits that app consumes, and the larger their monthly cost. This clearly leads apps that are more popular to build and integrate their own income system. Potentially opening up relationships between Spotify and the third party app for 'built in advertising' through the already in use web-api. Bands could advertise easily on third party apps through Spotify web-api to use potential third party apps that have a higher demographic of people who listen to that particular style of music.

There are to many of us who want to take an idea and combine it with music and another passion, writing software. While there's not enough reasons why this isn't a reality.

**Comment by arn95** [comment]
It's impossible to bring free music to hundreds or thousands through one premium account. I'm certain their API has query limitations just like any other API. You cross the quota they start to block the hose. If they don't I would be very surprised. And also, why would someone pay for a third party app when they have the source aka Spotify right there? I think that the point of Spotify APIs is to bring more users to the platform in any way possible... the apps that use the API alleviate the process or make it different, or fun, or whatever. Business to business will never work for Spotify in my opinion. Why do you think Apple is not using Spotify for music but instead made their own subscription service. It's probably more profitable to have loyal monthly subscribers...

**Comment by crwgregory** [comment]
The hole point of this issue is that there isn't a way to even play tracks using a third party api.

**Comment by GabeJacobs** [comment]
@crwgregory @arn95  Gotta weigh in here. We've waited for web playback for over a year now and have talked to Spotify about it in the past. First, there is no need for Spotify to do any sort of credit system. Having an full playback API for them is automatically good because it drives people to buy premium. It expands their market and allows for their brand to reach other parts of the internet.

That said, it's a security issue for them. It's also a manpower issue. As far as I know, only a few people work on the API's. We use their iOS SDK extensively and from my knowledge it does not seem like a true priority.

I hope Spotify sees this thread and understands just how much they could benefit for supporting full playback on the web. This thread is proof that their are developers who are waiting to build apps that spread Spotify. If there are developers, there are customers. That should be enough of a reason to prioritize this.

**Comment by JohnnyCrazy** [comment]
@GabeJacobs 
>That said, it's a security issue for them. It's also a manpower issue. As far as I know, only a few people work on the API's. We use their iOS SDK extensively and from my knowledge it does not seem like a true priority. In reality, their API is one of the few things that set them apart from their competitors and it could be the big difference that puts them above Apple Music, Tidal and Google Play.

While I aggree mostly with you, I still wish there would be some more transparency relating to future and current operations. Looking at the current issue list, there are about 44 enhancements, most of them with a statement like "maybe will come sometime in the future". Considering the rather small team and propably system-restrictions, I'm pretty sure a lot of those features won't come in the next year/s. What I would like to see: **A small analysis of the current open issues, considering implementability and a possible time-frame**.

>I hope Spotify sees this thread and understands just how much they could benefit for supporting full playback on the web. This thread is proof that their are developers who are waiting to build apps that spread Spotify. If there are developers, there are customers. That should be enough of a reason to prioritize this.

I'll second that first sentence, but considering the (dead) state of the `web-api`, `ios-sdk` and `android-sdk`-repo, I'm very uncertain.

(I still hope they're currently working on the desktop-SDK :wink: )

**Comment by rriemann** [comment]
I would love to see web podcasts being able to add music to their shows provided by Spotify (for an audience with Spotify Premium account of course). So many things would be possible. So please make it happen.

**Comment by spikespaz** [comment]
This needs to be added to the API. Where is `desktop-sdk`?

**Comment by gsugambit** [comment]
Was this ever implemented?

**Comment by jscholes** [comment]
> Was this ever implemented?

No.

**Comment by gordolio** [comment]
I'm following this issue until it is implemented. My project is "on hold" until it is. In the meantime I'm deciding if I want to spend the time creating an iPhone/Android app to support playback until this is implemented.

Overall the API is great! Can't wait to see this when it comes. I trust that spotify is working hard on creating this. There seems to be a fair amount of activity with the new WebAPI.

**Comment by osharifali** [comment]
@gordolio The Android SDK is terrible. You don't even have the ability to modify the MediaPlayer UI.

**Comment by Hackya** [comment]
Has anyone heard of SoundCloud? LOL

SoundCloud offers jQuery API which I don't really like (I would much more prefer a vanilla wrapped API) but it's extremely easy & quick to build your web app upon.

The only draw back of SoundCloud is that it currently has licenses with Universal Music & Warner Brother only.  Sony hasn't come on board yet.  But that's still roughly 50% of all major music licenses.

I wanted to created a Spotify web app but after reading this thread, I quickly realized I am better off staying with SoundCloud or host my own mp3.

**Comment by RareNCool** [comment]
@Hackya Yes. But they're two completely different things. If SoundCloud does the job for what you need that's great. But posturing it as a replacement for Spotify is like saying MSDN is a replacement for Stack Overflow; they're each good in their own way, but they are not the same thing.

**Comment by Hackya** [comment]
@RareNCool 

How can SoundCloud wep app be any different than Spotify web app other than the fact that Soptify web app does not allow full length streaming?

Can you demonstrate any test case/scenario where SoundCloud API is different from Spotify API?

Based on simple REST principles, our Web API endpoints return metadata in JSON format about artists, albums, and tracks directly from the Spotify catalogue. The API also provides access to user-related data such as playlists and music saved in a “Your Music” library, subject to user’s authorization. - sounds just like SoundCloud API to me.

https://developer.spotify.com/web-api/

Maybe there are other functionalities Spotify API provides that are not documented.

If so, please enlighten me.

**Comment by arn95** [comment]
@Hackya Didn't think of SoundCloud actually. Thanks! Before Spotify they were my goto for music. The only pro blem might be that it doesn't have the same amount of music as Spotify... We'll see.

**Comment by RareNCool** [comment]
@Hackya I'm not saying that the API isn't fundamentally the same, I'm saying that telling a Spotify subscriber who has cultivated their library, followings, etc. over the years, "No we can't do this with the Spotify API. But, hey! switch to SoundCloud and we can do it." isn't exactly a viable option because they simply aren't the same service. (and vice-versa for SoundCloud > Spotify) So unless you're building your app for new users only, offering one as an alternative to the other simply doesn't work.

**Comment by Hackya** [comment]
@RareNCool Oh, got you!!.  Thanks for clarifying what you mean.

**Comment by Hackya** [comment]
@arn95 Yes, that's why I looked into Spotify as well.  I guess we are all between a rock and a hard place. LOL

Either you get full length of roughly 50% of the major label songs or you get roughly 90% of the music but only 30 sec. of it. hahaha

My understanding is that SoundCloud is trying hard to have Sony come on board right now.

If that happens, SoundCloud will be able to offer same level of song selections as Spotify.  

So let's hope that 

1) SoundCloud can host Sony music and/or
2) Spotify offers full length streaming of music.

Either 1) or 2) will be great for many developers but if both 1) & 2) happens, that would simply be an awesomeness.

**Comment by mtimofiiv** [comment]
This would indeed be a really nice feature, to be able to play audio. I would assume this would be premium-only, so of course the problems with lost ad revenue and such would not be a valid argument.

If they share a virtual URI of the track, scoped to only that user, then it would make sense. You could also get a denial message if you try to access more than one of these virtual URIs at once, so the issue where two songs can be played simultaneously by the same account on two different devices would not be valid.

Another conceivable way to do this - you could create an audio stream for the premium user. You could then play a track by adding it into the stream's audio. Once again, it can be scoped to a single user and perhaps only one client can access it at a time.

My usecase is that I bought a Raspberry Pi, and I am working on a media centre type setup. I am so far missing Netflix and Spotify. The only way to get the setup to work is to use some convoluted audio stream bullcrap hooked up to a machine with the Spotify client installed, and that really sucks.

**Comment by jscholes** [comment]
> My usecase is that I bought a Raspberry Pi, and I am working on a media centre type setup. I am so far missing Netflix and Spotify.

As far as I know, you can still use [Mopidy](https://github.com/mopidy/mopidy) and [the Spotify add-on](https://github.com/mopidy/mopidy-spotify) to stream from Spotify on a Pi.  You can use an MPD client to search and control playback, although I'm not sure to what extent it supports playlist management etc.  Although the Web API does support pretty much everything else, so you can probably fill in any gaps yourself.

**Comment by marasco** [comment]
@thelinmichael today we haven't way to play full song via web api, true ? :(

**Comment by thelinmichael** [comment]
@marasco, that's correct.

**Comment by Jezza** [comment]
This is just fucking ridiculous.
This has been an issue for at least 2 years, and we've gotten literally nothing but "Yeah, this would be a nice thing."

First you remove any functionality with apps within spotify, without having full API support.
Next you moved away from a native UI to a crappy web UI that's just been wrapped in a program, that performs like shit on any computer.
And whenever someone says "Spotify has gone downhill", you're suddenly shocked.
Removing features, and not adding alternatives.
Making the program more shitty by moving away from a native UI.

Great, you're showing me a lot of consumer love here.
And the worse part is, it's been like this for ages.
A couple of weeks after you released the wrapped UI, I downgraded to 0.9.15.27, and have been there for a while, but I still use the latest version at work, and it's still fucking terrible.

If this doesn't get fixed soon, I'm just going to call it on Spotify.
I've been a consumer since 2011, and it's just gotten steadily worse.
Get your shit together.

**Comment by ptr2016** [comment]
So what's the point of the Spotify Web API, if there's no streaming for audio. Which is the entire point of Spotify, it's not like someone just wants to look at the titles of songs and albums.

**Comment by tokazio** [comment]
Using the linux lib for a while, i'm now testing the web api because the lib will be deprecated soon. It's cool, all is possible but playing tracks...

What's the interest of a audio service if it can't play tracks?!

What have you planned to replace the linux lib (used by a lot of embbed devices) ?

I don't want to work with android or ios.

**Comment by vincelwt** [comment]
Bonjour spotify/web-api,
Vincent L. vous a laissé un message sur IQ Elite 
Cliquez ici a.iqelite.com/j/QEVgD4/erg/?hg=yNZRrPTCtp pour lire votre message 


L'équipe d'IQ Elite



---------------------------------------------------------------------------
Ce message est destiné à reply@reply.github.com. Si vous ne souhaitez pas recevoir d'autres messages de la part de IQ Elite, veuillez cliquer sur ce lien : a.iqelite.com/common/unsubscribe/?code=b780882d&email=reply@reply.github.com.&hg=yNZRrPTCtp
IQ Elite est un service de IQ Elite S.L., dont le siège se trouve à : C/ Gran De Gràcia 15, 1a 1a, 08012 Barcelone, Espagne. 
---------------------------------------------------------------------------

**Comment by mtimofiiv** [comment]
Hey @vincelwt, looks like your email got hacked...

**Comment by aboe026** [comment]
I'm trying to create a Windows 10 UWP app for background music on Xbox One, since Spotify doesn't have a C# API I've resorted to the web API. This is a serious limitation and prevents any real Windows 10 UWP development. Seems stupid of Spotify to shoot themselves in the foot by preventing development on an entire ecosystem when they support others (Android/iOS).

**Comment by Jezza** [comment]
Hopefully we can get somewhere with this stupid issue.

https://twitter.com/SpotifyCares/status/798181185943781376

**Comment by mjunker** [comment]
+1

**Comment by kapral** [comment]
+1

**Comment by flaviolivolsi** [comment]
+1

**Comment by mtimofiiv** [comment]
Playback when?

May as well not even have an API if you can't make it do anything. That's like Github having an API that can't see any repos.

**Comment by diesphink** [comment]
+1

**Comment by oscarsiniscalchi** [comment]
+1

**Comment by tulustul** [comment]
:+1:

**Comment by glenndeb** [comment]
+1

**Comment by Aboisier** [comment]
+1

**Comment by GabeJacobs** [comment]
We've wanted this for years of course

**Comment by Cnordbo** [comment]
+1 
Vote for it on spotify idea suggestion https://community.spotify.com/t5/Idea-Submissions/All-Platforms-Other-API-Streaming-capabilities-for-full-track/idi-p/1559247

**Comment by lnhrdt** [comment]
Great idea @Cnordbo!

For everyone that winds up here in agreement, let's upvote @Cnordbo's Spotify community suggestion, add a 👍 to @Cnordbo's post and @func0der's original post at the top, and choose not to post a "+1" comment to cut down on the noise in here.

**Comment by jordanranson** [comment]
Spotify closed @Cnordbo's thread already, instead directing us to the developer site. Not helpful, @spotify :(

**Comment by jordanranson** [comment]
@spotify should either close this thread, saying they won't be adding this feature, or they should close this thread, saying that it will be added in the next API push. This endless run around is getting old.

**Comment by mtimofiiv** [comment]
Agree with @jakemason, it's the lack of communication that I think people find most frustrating.

**Comment by NicoHinderling** [comment]
+1 on @jordanranson.

Spotify, if you can't allow this feature for the Web API, at least address it in this thread and close the issue :pensive:

I can accept this not being available but at least acknowledge that all of us are in limbo as of right now.

**Comment by Jezza** [comment]
Holy fuck.
They just closed the request because "You might better be off trying your luck directly in the Spotify Dev community here: https://developer.spotify.com/"
Which is a link, rather indirectly, straight back to here.

I say indirectly, because it's just a landing page, but if you click support -> Bug Reports & Feature Requests -> Lo and behold, this exact repo.

What the actual fuck spotify...

**Comment by krak3n** [comment]
Switch to Google Play, for the same price you can stream tracks over HTTP, just knocked this up in `go`:

``` go
package main

import (
	"bufio"
	"fmt"
	"io"

	"github.com/budkin/gmusic"
	"github.com/korandiz/mpa"
	pulse "github.com/mesilliac/pulse-simple"
)

func main() {
	// Pulse Audio Stream
	ss := pulse.SampleSpec{pulse.SAMPLE_S16LE, 44100, 2}
	stream, err := pulse.Playback("sfm", "sfm", &ss)
	if err != nil {
		fmt.Println(err)
		return
	}
	defer stream.Free()
	defer stream.Drain()
	gm, err := gmusic.Login("foo", "bar")
	if err != nil {
		fmt.Println(err)
		return
	}
	rsp, err := gm.GetStream("4c1229c3-d82a-39b4-9b7e-b5184239abe3")
	if err != nil {
		fmt.Println(err)
		return
	}
	reader := bufio.NewReader(rsp.Body)
	fmt.Println("Content-Length:", rsp.ContentLength)
	fmt.Println("Status-Code:", rsp.StatusCode)
	// Data buffer
	data := make([]byte, 1024*8)
	// MPEG Decoder
	r := &mpa.Reader{Decoder: &mpa.Decoder{Input: reader}}
	// Play
	for {
		i, err := r.Read(data)
		if err == io.EOF || i == 0 {
			fmt.Println(err, i)
			return
		}
		i, err = stream.Write(data)
		if err != nil {
			fmt.Println("Can't write stream:", err)
			return
		}
	}
}
```

**Comment by Nearoo** [comment]
Well, Google Play doesn't have an official API at all, it seems.... Only an unofficial one, but that can download tracks at least.

**Comment by dicristomanuel** [comment]
+1

**Comment by noex** [comment]
+1

**Comment by BGrigorovich** [comment]
+1

**Comment by Swinkid** [comment]
+1

**Comment by chrisrossi** [comment]
The reason I dropped Spotify, FWIW, was because I couldn't integrate it with my Squeezebox.  If I'm going to pay $10/month, I at least should be able to listen to it in my living room.  With just this one feature added to the API, I'd have a shot at implementing it myself.

**Comment by dennerevaldt** [comment]
+1

**Comment by LRonHubs** [comment]
+1

**Comment by LRonHubs** [comment]
+69

**Comment by madgustoo** [comment]
+111111

**Comment by uniphonic** [comment]
+1

**Comment by Jezza** [comment]
People, please, stop leaving a "+1"

If you wish to show your desire for this ticket, use the reaction system.
That way you stop spamming everyone that's watching this issue.

So, before you add "+1", think about the 61 people that are going to see your "+1", and save us all getting our hopes up.

**Comment by LRonHubs** [comment]
That's a silly thing to say. Why wouldn't you direct your frustration at @spotify, who has left this issue unresolved for way too long?

**Comment by GabeJacobs** [comment]
Should we organize a massive all at once twitter war?????? Those work... right?

**Comment by LRonHubs** [comment]
Actually, I was going to send them a strongly worded letter. A bit old school I guess.

Here's what I have so far.

To Whom It May Concern:

It is not likely that I shall say anything new here. If I do, it will be of only minor significance. Nevertheless, I experienced quite an epiphany when I first realized that Spotify's metanarratives, though creative, are a yawning abyss of classism. Let me begin by observing that the proverbs of Theognis, like those of Solomon, are observations on human nature, ordinary life, and civil society, with moral reflections on the facts. I quote him as a witness of the fact that we've all heard Spotify yammer and whine about how it's being scapegoated again, the poor dear. In a similar vein, some day, in the far, far future, Spotify will realize that it is immovably entrenched in its contemptible philosophical positions. This realization will sink in slowly but surely and will be accompanied by a comprehension of how Spotify's perorations are not only passive-aggressive but divisive. They are divisive at a time when we need unity. They are prudish at a time when we need to come together to tell our shared stories about how Spotify holds onto power like the eunuch mandarins of the Forbidden City—sterile obstacles to progress who acquire power and use it to indoctrinate uncontrollable, carnaptious nymphomaniacs.

While there's no dispute that Spotify is gin-swilling and probably a little contumacious, it's also cunning, implacable, fanatic, and ruthless. Why else would it lower this country's moral tone and depreciate its commercial integrity? You know what I mean?

By trampling into the mud all that is fine and noble and beautiful, Spotify reveals its ignorance about ruffianism's polyvocality. It probably also doesn't realize that I want to make this clear so that those who do not understand deeper messages embedded within sarcastic irony—and you know who I'm referring to—can process my point. Spotify's sound bites are based on hate. Hate, interventionism, and an intolerance of another viewpoint, another way of life.

Spotify has a glib proficiency with words and very sensitive nostrils. It can smell money in your pocket from a block away. Once that delicious aroma reaches Spotify's nostrils, it'll start talking about the joy of antiheroism and how we should all bear the brunt of its actions. As you listen to Spotify's sing-song, chances are you won't even notice its hand as it goes into your pocket. Only later, after you realize you've been robbed, will you truly understand that no matter how bad you think its traducements are, I assure you that they are far, far worse than you think. Malicious, exploitative gits like Spotify are not born—they are excreted. However unsavory that metaphor may be, someone just showed me a memo supposedly written by Spotify. The memo spells out its plans to use cheap, intemperate propaganda to arouse the passions of the worst kinds of inficete rabble-rousers there are (also known as Spotify's compadres). If this memo is authentic, it tells us that Spotify's vituperations are eerily similar to those promoted by madmen such as Pol Pot. What's scary, though, is that their extollment of privatism has been ratcheted up a few notches from anything Pol Pot ever conjured up.

It seems to me that, as others have stated long before me, “I find Spotify the most pot-valiant organization in the world.” It is immature and stupid of Spotify to vilify our history, character, values, and traditions. It would be mature and intelligent, however, to open students' eyes, minds, hearts, and souls to the world around them, and that's why I say that every small-minded Svengali must be tempted at times to spit on his hands, hoist the black flag of snobbism, and begin to declare a national emergency, round up everyone who disagrees with it, and put them in concentration camps. That's sufficient evidence for me, at least, to conclude that Spotify's understrappers, while small in number, have powerful vocal chords. That conclusion is not based on some sort of despicable philosophy or on Spotify-style mental masturbation but on widely known and proven principles of science. These principles explain that mass anxiety is the equivalent of steroids for Spotify. If we feel helpless, Spotify is energized and ramps up its efforts to subject people to daily verbal, behavioral, and environmental indignities.

Spotify warrants that it is a paragon of morality and wisdom. Even if this were so, Spotify would still be surly. But Spotify should work with us, not step in at the eleventh hour and hog all the glory.

Although we can occasionally tie the retailers of paltry new claims to older fabrications, there is unfortunately no shortage of new rumor. If Spotify truly believes that once it has approved of something it can't possibly be imperious, then maybe it should enroll in Introduction to Reality 101. Spotify's crotchety advocates are nothing more than subservient blobs of easily controlled protoplasm. That's why they're so willing to help Spotify promote violence in all its forms—physical, sexual, psychological, economical, and social. Spotify demands that its snow jobs be discussed in only the most positive light. To ensure that this demand is met, it sends its fogyism outfit after anyone who fails to show the utmost deference when planting big, wet, sloppy kisses on Spotify's behind. Spotify has recently started distorting and trivializing the debate surrounding fascism. For some this development is a sign that a brave new world has arrived. For others it marks the beginning of the end of civilization. I lie in the second camp, primarily on the grounds that Spotify's grievances all stem from one, simple, faulty premise, that coercion in the name of liberty is a valid use of state power.

Spotify's squadristi were recently seen progressively enlarging and increasingly centralizing the means of oppression, exploitation, violence, and destruction. That's not a one-time accident or oversight. That's Spotify's policy. Mutual efforts against bestial, soporific boosterism are not just an educational process designed to teach people that it's time to step things up a notch and pronounce an enlightened and just judgment upon Spotify. These efforts also serve as a beacon, warning the world of the barbaric consequences of Spotify's heathenism-oriented, ludibrious anecdotes.

It is regrettable that Spotify's confederates have abdicated reason in this debate and allowed themselves to become captive of Spotify's lies, distortions, and hysteria, pure and simple. Just between you and me, when Spotify tells us that it is a master of precognition, psychokinesis, remote viewing, and other undeveloped human capabilities, it somehow fails to mention that its obiter dicta will come back to bite us in the behind quicker than you can double-check the spelling of “archaeopterygiformes”. It fails to mention that its stratagems are simply the result of vested interests striking back at a group whose actions in support of religious freedom, social reform, and government accountability have cut through those vested interests. And it fails to mention that it has found a way to avoid compliance with government regulations, circumvent any further litigation, and infringe upon our most important constitutional rights—all by trumping up a phony emergency. Spotify is firmly convinced that every featherless biped, regardless of intelligence, personal achievement, moral character, sense of responsibility, or sanity, should be given the power to violate strongly held principles regarding deferral of current satisfaction for long-term gains. Its belief is controverted, however, by the weight of the evidence indicating that Spotify adamantly maintains that it is a martyr for freedom and a victim of presenteeism. Such beliefs would be absolutely factual if it weren't for reality. As it stands, Spotify is a maladroit, putrid patrioteer. Consequently, attempting to respond to its crusades with logic is futile. A more productive response is to observe that Spotify's ruses are an unconscionable, blathering circle that begins and ends with Spotify. That's too big of a subject to get into here so let me instead discuss how dodgy, subhuman mouthpieces for invidious militarism are more susceptible to its brainwashing tactics than are any other group. Like water, their minds take the form of whatever receptacle Spotify puts them in. They then lose all recollection that like a verbal magician, Spotify knows how to lie without appearing to be lying, how to bury secrets in mountains of garbage-speak.

Spotify does not tolerate any view that differs from its own. Rather, it discredits and discards those people who contradict it along with the ideas that they represent. Spotify will fleece us long before it can convert me into one of its surrogates. Lastly, for those who read this letter, I unmistakably hope you take it to heart and pass this message on to others.

Sincerely,
L. Ron Hubbard's Thetan

**Comment by GabeJacobs** [comment]
@LRonHubs

![kevin-hart-wait-hello-who-are-you](https://cloud.githubusercontent.com/assets/953195/24681370/e184ef54-1962-11e7-920f-0bca42487c0b.jpg)

**Comment by hughrawlinson** [comment]
Hey everyone,

We're very much aware that web playback of full tracks is a really important feature of the platform, and we'd love to provide it as long as we're able to ensure the security of the artists content.

This issue seems to be a long list of 👍 s, and people getting annoyed at there being so many 👍 comments. I've locked this issue. Please use the react emoji on the opening comment to show support for this feature request.

Best,

Spotify Developer Support

**Comment by hughrawlinson** [comment]
Hi everyone!

I hope you'll be as excited as I am that I'm now able to close this issue following the announcement of our new [Web Playback SDK](https://beta.developer.spotify.com/documentation/web-playback-sdk/)! 🍾 🎆 🎉 🎈 🔈 🕸 

You'll find all the documentation on our [new website](https://beta.developer.spotify.com), where you can give feedback. You'll be able to file bugs as you encounter them, as well as feature requests, on our new [Web Playback SDK issue tracker](https://github.com/spotify/web-playback-sdk).

Happy hacking, we can't wait to see all the awesome things you build.

Best,

Hugh
Spotify for Developers

**Comment by TastesLikeTurkey** [comment]
@hughrawlinson Great news! Am I correct that this doesn't mean we'll be able to use the Web API to get a track? (e.g. playing a track from a WPF application).

**Comment by hughrawlinson** [comment]
Hi @TastesLikeTurkey - yes, that's correct, this SDK is for web playback.

**Comment by TastesLikeTurkey** [comment]
Understood. Is the issue really completely closed in that case, though? I may be wrong, but it was my impression that the primary interest here was playback via the Web _API_, not just on a web _platform_.

**Comment by TastesLikeTurkey** [comment]
@hughrawlinson To be clear, I'm not questioning the current implementation, just whether the issue ought to perhaps be left open. (Or does closing it mean that playback via the API is something Spotify isn't intersted in?)

**Comment by chrisrossi** [comment]
This is great for the browser, although Spotify already works fine in a browser, so I would imagine the uses cases would be somewhat limited.  My interest, and I'm guessing a lot of folk's interest, is in getting Spotify to run on devices that aren't officially supported.  I have a Roku and a Squeezebox in my living room and the Spotify app/plugin for those is unusable, so no Spotify in my living room.  With playback via the Web API, your community could expand the number of hardware platforms you can run on, which I imagine would only make your service more attractive to more people.

**Comment by jonknyc** [comment]
This is *awesome*.  Thank you Spotify for building this SDK!!!

**Comment by hughrawlinson** [comment]
@TastesLikeTurkey @chrisrossi we won't be able to provide access to raw audio data via the Web API for licensing reasons, but we're always interested in and working on ways to improve the platform, and we're specifically very aware of how much demand there is for playback across all kinds of devices. As I said above: "we'd love to provide it as long as we're able to ensure the security of the artists content", but I don't have anything to announce about it today.

**Comment by TastesLikeTurkey** [comment]
@hughrawlinson Fair enough. However, that basically puts us back to square one of this issue, so could it possibly remain open to track the original request (playback via the API)?

**Comment by hughrawlinson** [comment]
@TastesLikeTurkey there won't be playback via the API. The feature would likely happen as a separate SDK, and we're already acutely aware of the need for such an SDK. This issue is for "Web playback of Full Tracks", which is now possible 😄

**Comment by NikicaCh** [comment]
Can I provide the song's(tracks's) id and play the full song, not in my app but in the spotify player ?

**Comment by jonknyc** [comment]
Yep!

**Comment by NikicaCh** [comment]
And how ? I couldn't find that in the endpoints 🗡

**Comment by jonknyc** [comment]
First, find the device ID of your Spotify player using the devices endpoint:

https://developer.spotify.com/web-api/get-a-users-available-devices/

Then, use the Play endpoint, and specify the device_id parameter for your
Spotify player:

https://developer.spotify.com/web-api/start-a-users-playback/

-Jon

**Comment by NikicaCh** [comment]
And how to specify which song to play ?

**Comment by jonknyc** [comment]
With the uris parameter.

I think you can right click on a song in the Spotify player to find its track ID.
